pkg/runtime/runtime_test.go

```diff
@@ -5392,3 +5371,201 @@ func TestNamespacedPublisher(t *testing.T) {
 
 	assert.Equal(t, "ns1topic0", rt.pubSubs[TestPubsubName].component.(*mockPublishPubSub).PublishedRequest.Topic)
 }
+
+func TestGracefulShutdownPubSub(t *testing.T) {
+	rt := NewTestDaprRuntime(modes.StandaloneMode)
+	// setup pubsub
+	testGracefulRestartPubSub := "shutdownPubsub"
+	pubsubComponent := componentsV1alpha1.Component{
+		ObjectMeta: metaV1.ObjectMeta{
+			Name: testGracefulRestartPubSub,
+		},
+		Spec: componentsV1alpha1.ComponentSpec{
+			Type:     "pubsub.mockPubSub",
+			Version:  "v1",
+			Metadata: getFakeMetadataItems(),
+		},
+	}
+	rt.pubSubRegistry.RegisterComponent(
+		func(_ logger.Logger) pubsub.PubSub {
+			return &mockSubscribePubSub{}
+		},
+		"mockPubSub",
+	)
+
+	req := invokev1.NewInvokeMethodRequest("dapr/subscribe")
+	req.WithHTTPExtension(http.MethodGet, "")
+	req.WithRawData(nil, invokev1.JSONContentType)
+	subscriptionItems := []runtimePubsub.SubscriptionJSON{
+		{PubsubName: testGracefulRestartPubSub, Topic: "topic0", Route: "shutdown"},
+	}
+	sub, _ := json.Marshal(subscriptionItems)
+	fakeResp := invokev1.NewInvokeMethodResponse(200, "OK", nil)
+	fakeResp.WithRawData(sub, "application/json")
+
+	mockAppChannel := new(channelt.MockAppChannel)
+	rt.appChannel = mockAppChannel
+	mockAppChannel.On("InvokeMethod", mock.MatchedBy(matchContextInterface), req).Return(fakeResp, nil)
+
+	require.NoError(t, rt.initPubSub(pubsubComponent))
+	rt.startSubscriptions()
+	assert.NotNil(t, rt.pubsubCtx)
+	assert.NotNil(t, rt.topicCtxCancels)
+	assert.NotNil(t, rt.topicRoutes)
+
+	rt.runtimeConfig.GracefulShutdownDuration = 3 * time.Second
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		<-sigs
+		rt.Shutdown(rt.runtimeConfig.GracefulShutdownDuration)
+	}()
+
+	if p, err := os.FindProcess(os.Getpid()); err != nil {
+		rt.Shutdown(rt.runtimeConfig.GracefulShutdownDuration)
+	} else {
+		p.Signal(syscall.SIGTERM)
+	}
```

Comment:
can this just be `sigs <- syscall.SIGINT` or `sigs <- syscall.SIGTERM`? Also what piece of code is being tested here since rt.Shutdown() is being called manually anyway? Can we completely remove the signal processing logic in this test and call rt.Shutdown(3s) directly?

Similarly comment for other tests also below.

Reply:
The idea of this test is to init a SIGTERM and test. If we are unable to send sigterm we bypass the signal sending piece and call shutdown anyway. This logic is needed to ensure the test does not break in Windows. For Mac and Linux the test runs with a SIGTERM being sent to runtime.

Reply:
Removed SIGTERM and invoking Shutdown directly to mitigate flakiness of the tests in Windows environment